modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildController.java

```diff
@@ -0,0 +1,299 @@
+        return inBusyLockAsync(busyLock, () -> {
+            indexBuilder.stopBuildingIndexes(parameters.indexId());
+
+            return completedFuture(null);
+        });
+    }
+
+    private CompletableFuture<?> onPrimaryReplicaElected(PrimaryReplicaEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            TablePartitionId primaryReplicaId = (TablePartitionId) parameters.groupId();
+
+            if (isLocalNode(parameters.leaseholder())) {
+                primaryReplicaIds.add(primaryReplicaId);
+
+                // It is safe to get the latest version of the catalog because the PRIMARY_REPLICA_ELECTED event is handled on the
+                // metastore thread.
+                int catalogVersion = catalogService.latestCatalogVersion();
+
+                return getMvTableStorageFuture(parameters.causalityToken(), primaryReplicaId)
+                        .thenCompose(mvTableStorage -> awaitPrimaryReplicaFowNow(primaryReplicaId)
+                                .thenAccept(replicaMeta -> tryStartBuildIndexesForNewPrimaryReplica(
+                                        catalogVersion,
+                                        primaryReplicaId,
+                                        mvTableStorage,
+                                        replicaMeta
+                                ))
+                        );
+            } else {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return completedFuture(null);
+            }
+        });
+    }
+
+    private void tryStartBuildIndexesForNewPrimaryReplica(
+            int catalogVersion,
+            TablePartitionId primaryReplicaId,
+            MvTableStorage mvTableStorage,
+            ReplicaMeta replicaMeta
+    ) {
+        inBusyLock(busyLock, () -> {
+            if (isLeaseExpire(replicaMeta)) {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return;
+            }
+
+            // TODO: IGNITE-20530 We only need to get write-only indexes
+            for (CatalogIndexDescriptor indexDescriptor : catalogService.indexes(catalogVersion)) {
+                if (primaryReplicaId.tableId() == indexDescriptor.tableId()) {
+                    startBuildIndex(primaryReplicaId, indexDescriptor, mvTableStorage);
+                }
+            }
+        });
+    }
+
+    private void tryStartBuildIndex(
+            TablePartitionId primaryReplicaId,
+            CatalogIndexDescriptor indexDescriptor,
+            MvTableStorage mvTableStorage,
+            ReplicaMeta replicaMeta
+    ) {
+        inBusyLock(busyLock, () -> {
+            if (isLeaseExpire(replicaMeta)) {
+                stopBuildingIndexesIfPrimacyLost(primaryReplicaId);
+
+                return;
+            }
+
+            startBuildIndex(primaryReplicaId, indexDescriptor, mvTableStorage);
+        });
+    }
+
+    private void stopBuildingIndexesIfPrimacyLost(TablePartitionId replicaId) {
```

Comment:
I don’t quite understand what’s the typo here? There is such a word, here is the link: https://dictionary.cambridge.org/dictionary/english/primacy

Reply:
So you meant "primacy", not "primary". Ok, my bad in that case. I assumed that because there are no usages of the word "primacy" in the project.
Standard terminology would fit better in my opinion.
Anyway, why do we need a "if primacy lost", is the reason important to perform the operation?

Reply:
Which postfix do you think is better?
1. IfPrimacyExpire
2. IfPrimaryExpire

> Anyway, why do we need a "if primacy lost", is the reason important to perform the operation?

I did it by analogy with method `java.util.Map#putIfAbsent`, which hints that the operation will be performed only if the primary replica expired. First of all, it is convenient for me.

Reply:
It's more like "removeIfPresent", if I were to use your analogy. I would use `IfPrimaryExpire**d**`

Reply:
fix it